### PR TITLE
Fix NaN propagation behavior in torch.sign for CPU and CUDA backends

### LINKED ISSUE #186901: Inconsistent behavior of torch.sign on NaN input values
### 🐛 Describe the bug

There is a behavioral difference in how torch.sign handles NaN inputs between PyTorch and TensorFlow.
When feeding NaN values of float64 type to the sign operator:
TensorFlow 1.x returns 0 for NaN input.
After upgrading to TensorFlow 2.x, TensorFlow returns NaN for NaN input instead.
PyTorch currently returns 0 when processing NaN inputs with torch.sign.
There is an ambiguity about the standard expected output for torch.sign given NaN inputs: whether the result should be 0 or NaN. We would like to clarify and unify the official behavior.
```Python
import torch
import numpy as np

# Create float64 tensor with NaN
x = torch.tensor([np.nan], dtype=torch.float64)
out = torch.sign(x)
print(out)
```

### Versions

PyTorch version: 2.1.0
Python version: Python 3.8.19
OS: openEuler 22.03 LTS

### PR DESCRIPTION
**Pull request by Kinetic-Labs-GT**
### Description
This PR fixes an issue where `torch.sign` incorrectly handles `NaN` floating-point inputs by returning `0` instead of propagating the `NaN` value through the operator kernel. 

The current behavior violates modern numerical parity expectations (aligning with NumPy and TensorFlow 2.x) and standard IEEE 754 floating-point conventions where the sign of an undefined value should remain undefined (`NaN`).

Fixes #186901

### Technical Approach
The root cause was that both CPU and GPU kernels relied on relational comparison arithmetic (`(0 < a) - (a < 0)`). According to IEEE 754 specifications, any relational comparison against a `NaN` evaluates to `false`, which implicitly maps `NaN` inputs to a fallback `0` state.

1. **CPU Backend (`aten/src/ATen/native/cpu/UnaryOpsKernel.cpp`)**:
   - Upgraded the scalar lambda fallback to explicitly intercept `NaN` elements using `at::_isnan(a)` and return them directly.
   - Upgraded the vectorized SIMD path (`[=](Vectorized<scalar_t> self_vec)`) to compute a lane mask using `self_vec.isnan()` and apply a `Vectorized<scalar_t>::blendv` mask rewrite step across hardware registers to preserve original data lanes.
2. **CUDA Backend (`aten/src/ATen/native/cuda/UnarySignKernels.cu`)**:
   - Added an identical shortcut inside the device execution lambda (`[]GPU_LAMBDA`) to intercept elements via `at::_isnan(a)` before passing values down to `c10::signum(a)`.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01

### CODE REVIEW

**Comment by Kinetic-Labs-GT** [comment]
Hi @Aidyn-A, @eqy, and @syed-ahmed,

I have force-pushed an updated, compile-safe implementation to address the logical and architectural edge cases across both the CPU and CUDA backends.

### Core Changes & Fixes:

* **Compile-Time Branching (CPU & CUDA):** Added C++17 `if constexpr (std::is_integral_v<scalar_t>)` branching to decouple the type dispatch loops. This completely isolates the NaN-preservation logic to floating-point types (`float`, `double`, `Half`, `BFloat16`). As a result, integral type paths completely elide the `isnan` guards at compile time, eliminating template instantiation errors and preserving the highly optimized native fast paths.
* **CUDA Host-to-Device Fix:** Inside `sign_kernel_cuda`, the host-side `at::_isnan` utility inside the `GPU_LAMBDA` has been replaced with the device-safe qualified `c10::isnan` intrinsic. This resolves the NVCC `__host__` to `__device__` namespace violation completely.
* **Validation:** The updated files have been verified locally. Both `spin lint` and `lintrunner -a` pass with zero code style or formatting errors, and local target compilation succeeds perfectly.

The PR is now fully ready. Please take a look when you have a moment!
